app/uk/gov/hmrc/agentclientrelationships/controllers/Relationships.scala

```diff
@@ -74,42 +78,54 @@ class Relationships @Inject()(service: RelationshipsService) extends BaseControl
-    implicit val auditData = new AuditData()
-    auditData.set("arn", arn)
+  def create(arn: Arn, mtdItId: MtdItId): Action[AnyContent] = agentOrClient.async { implicit request =>
+    forThisAgentOrClient(arn, mtdItId) {
+      implicit val auditData = new AuditData()
+      auditData.set("arn", arn)
 
-    (for {
-      agentCode <- service.getAgentCodeFor(arn)
-      _ <- service.checkForRelationship(mtdItId, agentCode)
-        .map(_ => throw new Exception("RELATIONSHIP_ALREADY_EXISTS"))
+      (for {
+        agentCode <- service.getAgentCodeFor(arn)
+        _ <- service.checkForRelationship(mtdItId, agentCode)
+          .map(_ => throw new Exception("RELATIONSHIP_ALREADY_EXISTS"))
+          .recover {
+            case RelationshipNotFound("RELATIONSHIP_NOT_FOUND") => ()
+          }
+        _ <- service.createRelationship(arn, mtdItId, Future.successful(agentCode), Set(), false, true)
+      } yield ())
+        .map(_ => Created)
         .recover {
-          case RelationshipNotFound("RELATIONSHIP_NOT_FOUND") => ()
+          case ex =>
+            Logger.warn(s"Could not create relationship: ${ex.getMessage}")
+            NotFound(toJson(ex.getMessage))
         }
-      _ <- service.createRelationship(arn, mtdItId, Future.successful(agentCode), Set(), false, true)
-    } yield ())
-      .map(_ => Created)
-      .recover {
-        case ex =>
-          Logger.warn(s"Could not create relationship: ${ex.getMessage}")
-          NotFound(toJson(ex.getMessage))
-      }
+    }
   }
 
-  def delete(arn: Arn, mtdItId: MtdItId): Action[AnyContent] = Action.async { implicit request =>
-    implicit val auditData = new AuditData()
-    auditData.set("arn", arn)
+  def delete(arn: Arn, mtdItId: MtdItId): Action[AnyContent] = agentOrClient.async { implicit request =>
+    forThisAgentOrClient(arn, mtdItId) {
+      implicit val auditData = new AuditData()
+      auditData.set("arn", arn)
+
+      (for {
+        agentCode <- service.getAgentCodeFor(arn)
+        _ <- service.checkForRelationship(mtdItId, agentCode)
+        _ <- service.deleteRelationship(arn, mtdItId)
+      } yield ())
+        .map(_ => NoContent)
+        .recover {
+          case ex =>
+            Logger.warn(s"Could not delete relationship: ${ex.getMessage}")
+            NotFound(toJson(ex.getMessage))
+        }
+    }
+  }
 
-    (for {
-      agentCode <- service.getAgentCodeFor(arn)
-      _ <- service.checkForRelationship(mtdItId, agentCode)
-      _ <- service.deleteRelationship(arn, mtdItId)
-    } yield ())
-      .map(_ => NoContent)
-      .recover {
-        case ex =>
-          Logger.warn(s"Could not delete relationship: ${ex.getMessage}")
-          NotFound(toJson(ex.getMessage))
-      }
+  private def forThisAgentOrClient(arn: Arn, mtdItId: MtdItId)
+                                  (block: => Future[Result])(implicit request: AgentOrClient[_]) = {
+    (arn != request.arn.getOrElse(""), mtdItId != request.mtdItId.getOrElse("")) match {
```

Comment:
@cavster please can you re-review, if the whole PR looks good to you please merge